File: client/artifact_app/lib/models/schedule.dart
class Schedule {
  final String id; // VARCHAR(6)
  final String artifactId; // VARCHAR(6)
  final String? artifactName;
  final DateTime scheduledDate;
  final String scheduledTime;
  final String operatorUsername;
  final String notes;
  final bool completed;
  final DateTime createdAt;

  Schedule({
    required this.id,
    required this.artifactId,
    this.artifactName,
    required this.scheduledDate,
    required this.scheduledTime,
    required this.operatorUsername,
    required this.notes,
    required this.completed,
    required this.createdAt,
  });

  factory Schedule.fromJson(Map<String, dynamic> json) {
    return Schedule(
      id: json['id']?.toString() ?? '',
      artifactId: json['artifact_id']?.toString() ?? '',
      artifactName: json['artifact_name'] as String?,
      scheduledDate:
          DateTime.tryParse(json['scheduled_date']?.toString() ?? '') ??
              DateTime.now(),
      scheduledTime: json['scheduled_time']?.toString() ?? '09:00',
      operatorUsername: json['operator_username']?.toString() ?? '',
      notes: json['notes']?.toString() ?? '',
      completed: json['completed'] as bool? ?? false,
      createdAt:
          DateTime.tryParse(json['created_at']?.toString() ?? '') ?? DateTime.now(),
    );
  }
}

File: server/app/schemas/schedule.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class ScheduleBase(BaseModel):
    artifact_id: int
    scheduled_date: datetime
    scheduled_time: str = Field(default="09:00", pattern=r"^\d{2}:\d{2}$")
    operator_username: str = Field(default="", max_length=100)
    notes: str = Field(default="", max_length=2000)


class ScheduleCreate(ScheduleBase):
    pass


class ScheduleUpdate(BaseModel):
    scheduled_date: datetime | None = None
    scheduled_time: str | None = Field(default=None, pattern=r"^\d{2}:\d{2}$")
    operator_username: str | None = Field(default=None, max_length=100)
    notes: str | None = Field(default=None, max_length=2000)
    completed: bool | None = None


class ScheduleRead(ScheduleBase):
    model_config = ConfigDict(from_attributes=True)

    id: int
    completed: bool
    created_at: datetime
    artifact_name: str | None = None
